File: graph/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (numCourses < 0 || prerequisites == null) return false;
        Map<Integer, List<Integer>> graph = new HashMap<>();
        Deque<Integer> queue = new ArrayDeque<>();
        int[] indegree = new int[numCourses];
        Arrays.fill(indegree, 0);
        int ans = 0;

        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new ArrayList<>());
        }

        for (int[] preq : prerequisites) {
            int src = preq[1];
            int dest = preq[0];
            graph.get(src).add(dest);
            indegree[dest] += 1;
        }
        
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            int node = queue.pollFirst();
            ans += 1;

            for (int nei : graph.get(node)) {
                indegree[nei] -= 1;
                if (indegree[nei] == 0) {
                    queue.offer(nei);
                }
            }

        }

        return ans == numCourses;
    }
}

File: graph/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        indegree = [0] * numCourses
        queue = []
        count = 0

        for dest, src in prerequisites:
            graph[src].append(dest)
            indegree[dest] += 1

        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        while queue:
            node = queue.pop(0)
            count += 1

            for nei in graph[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    queue.append(nei)
        
        return count == numCourses

        
